### PR TITLE
Fix PostgreSQL `include` indexes with keyword column names

### LINKED ISSUE #51428: Problem with field named timestamp and index with include
### Steps to reproduce

When you have a field named "timestamp" and want to add an index including this field, the schema is generated with escape characters, making it impossible to run the load schema

```ruby
class ChangeIndexStatusToEvent < ActiveRecord::Migration[7.1]
  def change
    add_index :events, [:status, :name, :app_id], name: 'index_events_on_status_name_app_id', include: [:id, :timestamp]
  end
end
```

This schema is generated when running the migrate above:

```ruby
t.index ["status", "name", "app_id", "timestamp"], name: "index_events_on_status_name_app_id", include: ["id", "\"timestamp\""]
```

Test case:

```ruby
# frozen_string_literal: true

require 'bundler/inline'

gemfile(true) do
  source 'https://rubygems.org'

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem 'rails'

  gem 'pg'
end

require 'active_record'
require 'minitest/autorun'
require 'logger'

# `CREATE DATABASE rails_51428_test;`
ActiveRecord::Base.establish_connection(adapter: 'postgresql', database: 'rails_51428_test')
ActiveRecord::Base.logger = Logger.new($stdout)

ActiveRecord::Schema.define do
  create_table :events, force: true do |t|
    t.decimal :amount, precision: 10, scale: 0, default: 0, null: false
    t.datetime :timestamp
  end

  add_index :events, [:amount], name: 'index_events_on_timestamp', include: [:timestamp]
end

ActiveRecord::SchemaDumper.dump

class Event < ActiveRecord::Base; end

class EventTest < Minitest::Test
  def test_stuff
    assert_equal 0, Event.count
    Event.create!(amount: 1)
    assert_equal 1, Event.count

    # Note: this fails silently. I later discovered that neither this nor
    # `ActiveRecord::Tasks::DatabaseTasks.load_schema_current(:ruby, ENV['SCHEMA'])`
    # fail. See comments below.
    puts `rails db:schema:load`
  end
end
```

I am not sure if the word "timestamp" is reserved for the system, but I started having this problem after upgrading the rails version from 7.0.3 to 7.1.0

### Expected behavior
Don't escape the characters at include in index or add this explanation to take care to use field named "timestamp" in Rails Guide.

### Actual behavior
Using field named "timestamp" and adding index with this field in include, results in the schema with escaped characters, making it impossible to run the load schema.

### System configuration
**Rails version**: 7.1.0

**Ruby version**: 3.3.0

### PR DESCRIPTION
**Pull request by fatkodima**
Fixes #51428.

### CODE REVIEW

**Comment by JoeDupuis** [comment]
I wonder if that's really the correct fix here :thinking: 
This fix would work for the original case, but it would fail for more complex column names. 

I think anyone doing this is looking for trouble, but nothing would prevent someone from having quotes inside their column names. 

This is not likely to happen with columns created by humans, but I've seen it happen on columns created automatically by various systems.

Take this repro script: 

```ruby
# frozen_string_literal: true

require 'bundler/inline'

gemfile(true) do
  source 'https://rubygems.org'

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }
  if ENV['custom_rails_path']
    gem 'rails', path: ENV['custom_rails_path']
  else
    gem 'rails'
  end

  gem 'pg'
  gem 'debug'
end

require 'active_record'
require 'minitest/autorun'
require 'logger'
require 'debug'

# `CREATE DATABASE rails_51428_test;`
database = 'rails_51428_test'
%x( dropdb --if-exists #{database} )
%x( createdb -E UTF8 -T template0 #{database} --lc-collate en_US.UTF-8)
ActiveRecord::Base.establish_connection(adapter: 'postgresql', database:)
ActiveRecord::Base.logger = Logger.new($stdout)

ActiveRecord::Schema.define do
  create_table :events, force: true do |t|
    t.decimal :amount, precision: 10, scale: 0, default: 0, null: false
    t.datetime :timestamp
    t.datetime :some_name
    t.datetime :"some other name"
    t.datetime :"I\"like\"pain"
    t.datetime :bigint
  end

  add_index :events, [:amount], name: 'index_events_on_timestamp1', include: ["some_name"]
  add_index :events, [:amount], name: 'index_events_on_timestamp2', include: ["some other name"]
  add_index :events, [:amount], name: 'index_events_on_timestamp3', include: ["timestamp"]
  add_index :events, [:amount], name: 'index_events_on_timestamp4', include: ["bigint"]
  add_index :events, [:amount], name: 'index_events_on_timestamp5', include: ["I\"like\"pain"]
end

ActiveRecord::SchemaDumper.dump

class Event < ActiveRecord::Base; end

class EventTest < Minitest::Test
  def test_stuff
  end
end

```

It  would dump these indexes:
```
    t.index ["amount", "I\"like\"pain"], name: "index_events_on_timestamp5", include: ["\"I\"\"like\"\"pain\""]
    t.index ["amount", "bigint"], name: "index_events_on_timestamp4", include: ["\"bigint\""]
    t.index ["amount", "some other name"], name: "index_events_on_timestamp2", include: ["\"some other name\""]
    t.index ["amount", "timestamp"], name: "index_events_on_timestamp3", include: ["\"timestamp\""]
    t.index ["amount"], name: "index_events_on_timestamp1", include: ["some_name"]
```
  
 When querying pg directly:
 ```
    "index_events_on_timestamp1" btree (amount) INCLUDE (some_name)
    "index_events_on_timestamp2" btree (amount) INCLUDE ("some other name")
    "index_events_on_timestamp3" btree (amount) INCLUDE ("timestamp")
    "index_events_on_timestamp4" btree (amount) INCLUDE ("bigint")
    "index_events_on_timestamp5" btree (amount) INCLUDE ("I""like""pain")
```
Notice PG escaping anything weird like reserved terms, spaces or double quotes (by doubling the double quotes).

I am thinking, perhaps we can "parse" the string from PG format. This would eliminate the surrounding quotes and take care of de-doubling/escaping any real double quote in column names.

I am still looking if there is anything already present in Rails to parse the sql/pg string format. 
I don't think it's desirable to pull a parser or to write one. 
That's assuming the string format is not super simple. If the only difference is converting `""` to  `\"` and eliminating single double quotes perhaps we could just do it in place.

File: activerecord/lib/active_record/connection_adapters/postgresql/schema_statements.rb [hunk 1]
```diff
@@ -112,7 +112,7 @@ def indexes(table_name) # :nodoc:
 
             orders = {}
             opclasses = {}
-            include_columns = include ? include.split(",").map(&:strip) : []
+            include_columns = include ? include.split(",").map { |c| c.strip.delete('"') } : []
```
Comment (JoeDupuis) [inline activerecord/lib/active_record/connection_adapters/postgresql/schema_statements.rb:115]: There is already a helper to remove the surrounding quote.

I also added something to convert the double-double-quotes.

```suggestion
            include_columns = include ? include.split(",").map {|c| Utils.unquote_identifier(c.strip.gsub('""', '"'))} : []
```

**Review by JoeDupuis — COMMENTED** [review]
[Regarding my previous comment](https://github.com/rails/rails/pull/51429#issuecomment-2021739005)

Turns out there is already a helper method for removing the surrounding quotes without touching the inside ones. 

I added a suggestion.

I tested the suggestion inside a rails app and the output of the schema matches the migration + `bin/rails db:schema:load` is executing without crashing.

I _think_ this would be enough. I tested other special characters (`"%'â€™`). The output matches the input and it doesn't crash schema load.

File: activerecord/test/cases/adapters/postgresql/postgresql_adapter_test.rb [hunk 2]
```diff
@@ -346,6 +346,14 @@ def test_include_multiple_columns_index
...
         end
+
+        def test_include_keyword_column_name
+          with_example_table("id integer, timestamp integer") do
+            @connection.add_index "ex", :id, name: "include", include: [:timestamp]
+            index = @connection.indexes("ex").find { |idx| idx.name == "include" }
+            assert_equal ["timestamp"], index.include
+          end
+        end
       end
```
Comment (JoeDupuis) [inline activerecord/test/cases/adapters/postgresql/postgresql_adapter_test.rb:358]: Could be a good idea to add a regression test for quotes in column names:

```suggestion

      def test_include_escaped_quotes_column_name
        with_example_table(%{id integer, "I""like""quotes" integer}) do
          @connection.add_index "ex", :id, name: "include", include: [:"I\"like\"quotes"]
          index = @connection.indexes("ex").find { |idx| idx.name == "include" }
          assert_equal ["I\"like\"quotes"], index.include
        end
      end
```

File: activerecord/lib/active_record/connection_adapters/postgresql/schema_statements.rb (hunk 1 above)
Comment (fatkodima) [reply activerecord/lib/active_record/connection_adapters/postgresql/schema_statements.rb]: Thanks 👍, updated! I hope people do not use (at least explicitly) quotes in their column or table names 😆

**Review by JoeDupuis — COMMENTED** [review]
🙌 :shipit: 


> Thanks 👍, updated! I hope people do not use (at least explicitly) quotes in their column or table names 😆

Oh, I hope too 😆 I think it's a pretty niche issue (systems generating columns), but it's still worth guarding against since it's not more difficult to implement.

I've seen beginners learning programming try to use spaces instead of underscores though. 😮 

Now that we are talking about it, this makes me want to to hide an emoji in a column name somewhere 🤔

**Comment by fatkodima** [comment]
@yahonda Can you also please take a look at this PR?

**Pull request merged**